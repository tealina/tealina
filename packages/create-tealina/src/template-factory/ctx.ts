export interface CtxForMakeCode {
  isRestful: boolean
  framwork: 'express' | 'fastify'
}
export interface TemplateSnap {
  filename: string
  code: string
}

type SupportFramworks = 'express' | 'fastify'

export const replyExpression: Record<SupportFramworks, string> = {
  express: "'  res.sendStatus(200)',",
  fastify: "'  res.send()',",
}
