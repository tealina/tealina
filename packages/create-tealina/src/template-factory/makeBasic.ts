import { CtxForMakeCode, replyExpression } from './ctx'

export const makeBasicCode = (ctx: CtxForMakeCode) => {
  return [
    'export default makeTemplate(({ relative2api }) =>',
    '  [',
    "    `import type { AuthedHandler } from '${relative2api}/../types/handler.js'`,",
    "    `import { convention } from '${relative2api}/convention.js'`,",
    "    '',",
    '    `/** TODO: describe what it does */`,',
    '    `const handler: AuthedHandler = async (req, res) => {`,',
    `     ${replyExpression[ctx.framwork]}`,
    "    '}',",
    "    '',",
    '    `export default convention(handler)`,',
    "    '',",
    "  ].join('\\n')",
    ')',
  ].join('\n')
}
