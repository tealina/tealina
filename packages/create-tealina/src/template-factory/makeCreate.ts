import { CtxForMakeCode } from './ctx'

const restfulStyle = {
  fn: [
    'export default makeTemplate(',
    '  ({ Filename: Model, relative2api, filename: model }) => {',
  ],
}
const postGetStyle = {
  fn: [
    'export default makeTemplate(({ Dir: Model, relative2api, dir: model }) => {',
  ],
}

export const makeCreationCode = (ctx: CtxForMakeCode) => {
  const actual = ctx.isRestful ? restfulStyle : postGetStyle
  return [
    actual.fn,
    '  const imps = [',
    "    `import type { AuthedHandler } from '${relative2api}/../types/handler.js'`,",
    "    `import type { Pure } from '${relative2api}/../types/pure.js'`,",
    "    `import { convention } from '${relative2api}/convention.js'`,",
    "    `import { db } from '${relative2api}/db/prisma.js'`,",
    '  ]',
    '  const codes = [',
    '    `type ApiType = AuthedHandler<{ body: Pure.${Model}CreateInput }, Pure.${Model}>`,',
    "    '',",
    '    `/** Create ${Model} */`,',
    '    `const handler: ApiType = async (req, res) => {`,',
    '    `  const result = await db.${model}.create({`,',
    "    '    data: req.body,',",
    "    '  })',",
    "    '  res.send(result)',",
    "    '}',",
    "    '',",
    '    `export default convention(handler)`,',
    "    '',",
    '  ]',
    "  return [...imps, '', ...codes].join('\\n')",
    '})',
  ]
    .flat()
    .join('\n')
}
