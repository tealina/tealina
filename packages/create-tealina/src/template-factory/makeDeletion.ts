import { CtxForMakeCode, replyExpression } from './ctx'
const restfulStyle = {
  imps: [
    "    `import type { RawId } from '${relative2api}/../types/common.js'`,",
    "    `import { modelIdZ } from '${relative2api}/validate/modelId.js'`,",
  ],
  apiType: ['  `type ApiType = AuthedHandler<{ params: RawId }>`,'],

  body: ["    '  const { id } = modelIdZ.parse(req.params)',"],
}
const postGetStyle = {
  imps: [
    "    `import type { ModelId } from '${relative2api}/../types/common.js'`,",
  ],
  apiType: ['    `type ApiType = AuthedHandler<{ body: ModelId }>`,'],
  body: "    '  const { id } = req.body',",
}
export const makeDeletetionCode = (ctx: CtxForMakeCode) => {
  const actual = ctx.isRestful ? restfulStyle : postGetStyle
  return [
    'export default makeTemplate(({ relative2api, Dir: Model, dir: model }) => {',
    '  const imps = [',
    "    `import type { AuthedHandler } from '${relative2api}/../types/handler.js'`,",
    actual.imps,
    "    `import { convention } from '${relative2api}/convention.js'`,",
    "    `import { db } from '${relative2api}/db/prisma.js'`,",
    '  ]',
    '  const codes = [',
    actual.apiType,
    "    '',",
    '    `/** Delete ${Model} by id */`,',
    '    `const handler: ApiType = async (req, res) => {`,',
    actual.body,
    '    `  await db.${model}.delete({ where: { id } })`,',
    `    ${replyExpression[ctx.framwork]}`,
    "    '}',",
    "    '',",
    '    `export default convention(handler)`,',
    "    '',",
    '  ]',
    "  return [...imps, '', ...codes].join('\\n')",
    '})',
  ]
    .flat()
    .join('\n')
}
