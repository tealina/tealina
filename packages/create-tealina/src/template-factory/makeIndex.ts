import { CtxForMakeCode } from './ctx'

const imps = [
  '// @ts-check',
  "import genCreateCode from './genCreateCode.mjs'",
  "import genUpdateCode from './genUpdateCode.mjs'",
  "import genDeleteCode from './genDeleteCode.mjs'",
  "import genReadCode from './genReadCode.mjs'",
  "import genBasicCode from './genBasicCode.mjs'",
  "import { defineApiTemplates } from 'tealina'",
  '',
]

const restfulStyle = {
  create: ["    name: '',", "    method: 'post',"],
  read: ["    name: '',", "    method: 'get',"],
  update: ["    name: '[id]',", "    method: 'put',"],
  dele: ["    name: '[id]',", "    method: 'delete',"],
}
const postGetStyle = {
  create: ["    name: 'create',", "    method: 'post',"],
  read: ["    name: 'getList',", "    method: 'post',"],
  update: ["    name: 'update',", "    method: 'post',"],
  dele: ["    name: 'delete',", "    method: 'post',"],
}

export const makeIndexCode = (ctx: CtxForMakeCode) => {
  const actual = ctx.isRestful ? restfulStyle : postGetStyle
  return [
    imps,
    'export default defineApiTemplates([',
    '  {',
    "    alias: 'c',",
    actual.create,
    '    generateFn: genCreateCode,',
    '  },',
    '  {',
    "    alias: 'r',",
    actual.read,
    '    generateFn: genReadCode,',
    '  },',
    '  {',
    "    alias: 'u',",
    actual.update,
    '    generateFn: genUpdateCode,',
    '  },',
    '  {',
    "    alias: 'd',",
    actual.dele,
    '    generateFn: genDeleteCode,',
    '  },',
    '  {',
    "    alias: '*', //fallback",
    "    name: '',",
    "    method: 'post',",
    '    generateFn: genBasicCode,',
    '  },',
    '])',
  ]
    .flat()
    .join('\n')
}
