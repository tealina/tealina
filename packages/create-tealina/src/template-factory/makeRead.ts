import { CtxForMakeCode } from './ctx'

const restfulStyle = {
  fn: [
    'export default makeTemplate(',
    '  ({ Filename: Model, relative2api, filename: model }) => {',
  ],
  imps: [
    "    `import { findManyArgsZ } from '${relative2api}/validate/findManyArgs.js'`,",
    "    `import type { PageResult, RawFindManyArgs } from '${relative2api}/../types/common.js'`,",
  ],
  apiType:
    '`type ApiType = AuthedHandler<{ query: RawFindManyArgs }, PageResult<Pure.${Model}>>`,',
  query: '`  const findManyArgs = findManyArgsZ.parse(req.query)`,',
}
const postGetStyle = {
  fn: [
    'export default makeTemplate(({ Dir: Model, relative2api, dir: model }) => {',
  ],
  imps: [
    "    `import type { FindManyArgs, PageResult } from '${relative2api}/../types/common.js'`,",
  ],
  apiType:
    '`type ApiType = AuthedHandler<{ body: FindManyArgs }, PageResult<Pure.${Model}>>`,',
  query: '`  const findManyArgs = req.body`,',
}

export const makeReadCode = (ctx: CtxForMakeCode) => {
  const actual = ctx.isRestful ? restfulStyle : postGetStyle
  return [
    actual.fn,
    '  const imps = [',
    "    `import type { AuthedHandler } from '${relative2api}/../types/handler.js'`,",
    "    `import type { Pure } from '${relative2api}/../types/pure.js'`,",
    actual.imps,
    "    `import { convention } from '${relative2api}/convention.js'`,",
    "    `import { db } from '${relative2api}/db/prisma.js'`,",
    '  ]',
    '  const codes = [',
    `    ${actual.apiType}`,
    "    '',",
    '    `/** Get page datas from ${Model} */`,',
    '    `const handler: ApiType = async (req, res) => {`,',
    `    ${actual.query}`,
    "    '  const [total, datas] = await db.$transaction([',",
    '    `    db.${model}.count({ where: findManyArgs.where }),`,',
    '    `    db.${model}.findMany(findManyArgs),`,',
    "    '  ])',",
    "    '  res.send({ total, datas })',",
    "    '}',",
    "    '',",
    '    `export default convention(handler)`,',
    "    '',",
    '  ]',
    "  return [...imps, '', ...codes].join('\\n')",
    '})',
  ]
    .flat()
    .join('\n')
}
