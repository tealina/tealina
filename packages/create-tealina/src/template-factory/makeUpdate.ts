import { CtxForMakeCode } from './ctx'

const restfulStyle = {
  imps: [
    "    `import type { RawId } from '${relative2api}/../types/common.js'`,",
    "    `import { modelIdZ } from '${relative2api}/validate/modelId.js'`,",
  ],
  apiType: [
    '  `type ApiType = AuthedHandler<{ params: RawId; body: Pure.${Model}UpdateInput }>`,',
  ],

  body: [
    "    '  const { id } = modelIdZ.parse(req.params)',",
    "    '  const data = req.body',",
  ],
}
const postGetStyle = {
  imps: [
    "    `import type { ModelId } from '${relative2api}/../types/common.js'`,",
  ],
  apiType: [
    '    `type ApiType = AuthedHandler<`,',
    '    `  { body: ModelId & Pure.${Model}UpdateInput },`,',
    '    `  Pure.${Model}`,',
    '    `>`,',
  ],
  body: "    '  const { id, ...data } = req.body',",
}

export const makeUpdateCode = (ctx: CtxForMakeCode) => {
  const actual = ctx.isRestful ? restfulStyle : postGetStyle
  return [
    'export default makeTemplate(({ Dir: Model, relative2api, dir: model }) => {',
    '  const imps = [',
    "    `import type { AuthedHandler } from '${relative2api}/../types/handler.js'`,",
    "    `import type { Pure } from '${relative2api}/../types/pure.js'`,",
    "    `import { convention } from '${relative2api}/convention.js'`,",
    "    `import { db } from '${relative2api}/db/prisma.js'`,",
    actual.imps,
    '  ]',
    '  const codes = [',
    actual.apiType,
    "    '',",
    '    `/** Update ${Model} by id */`,',
    '    `const handler: ApiType = async (req, res) => {`,',
    actual.body,
    '    `  const result = await db.${model}.update({`,',
    "    '    where: { id },',",
    "    '    data,',",
    "    '  })',",
    "    '  res.send(result)',",
    "    '}',",
    "    '',",
    '    `export default convention(handler)`,',
    "    '',",
    '  ]',
    "  return [...imps, '', ...codes].join('\\n')",
    '})',
  ]
    .flat()
    .join('\n')
}
