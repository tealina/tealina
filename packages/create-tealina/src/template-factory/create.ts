import { CtxForMakeCode, TemplateSnap } from './ctx'
import { makeBasicCode } from './makeBasic'
import { makeCreationCode } from './makeCreate'
import { makeDeletetionCode } from './makeDeletion'
import { makeIndexCode } from './makeIndex'
import { makeReadCode } from './makeRead'
import { makeUpdateCode } from './makeUpdate'

export const createTemplates = (ctx: CtxForMakeCode): TemplateSnap[] => {
  return [
    {
      filename: 'genBasicCode.mjs',
      code: makeBasicCode(ctx),
    },
    { filename: 'genCreateCode.mjs', code: makeCreationCode(ctx) },
    { filename: 'genReadCode.mjs', code: makeReadCode(ctx) },
    { filename: 'genUpdateCode.mjs', code: makeUpdateCode(ctx) },
    { filename: 'genDeleteCode.mjs', code: makeDeletetionCode(ctx) },
    { filename: 'index.mjs', code: makeIndexCode(ctx) },
  ]
}
