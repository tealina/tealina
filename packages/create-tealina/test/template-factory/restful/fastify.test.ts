import { expect, test } from 'vitest'
import { createTemplates } from '../../../src/template-factory/create'

test('make fastify restful templates', async () => {
  const snpas = createTemplates({
    isRestful: true,
    framwork: 'fastify',
  })
  expect(snpas).toMatchSnapshot()
})
