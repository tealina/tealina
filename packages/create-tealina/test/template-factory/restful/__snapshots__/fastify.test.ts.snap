// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`make fastify restful templates 1`] = `
[
  {
    "code": "export default makeTemplate(({ relative2api }) =>
  [
    \`import type { AuthedHandler } from '\${relative2api}/../types/handler.js'\`,
    \`import { convention } from '\${relative2api}/convention.js'\`,
    '',
    \`/** TODO: describe what it does */\`,
    \`const handler: AuthedHandler = async (req, res) => {\`,
     '  res.send()',
    '}',
    '',
    \`export default convention(handler)\`,
    '',
  ].join('\\\\n')
)",
    "filename": "genBasicCode.mjs",
  },
  {
    "code": "export default makeTemplate(
  ({ Filename: Model, relative2api, filename: model }) => {
  const imps = [
    \`import type { AuthedHandler } from '\${relative2api}/../types/handler.js'\`,
    \`import type { Pure } from '\${relative2api}/../types/pure.js'\`,
    \`import { convention } from '\${relative2api}/convention.js'\`,
    \`import { db } from '\${relative2api}/db/prisma.js'\`,
  ]
  const codes = [
    \`type ApiType = AuthedHandler<{ body: Pure.\${Model}CreateInput }, Pure.\${Model}>\`,
    '',
    \`/** Create \${Model} */\`,
    \`const handler: ApiType = async (req, res) => {\`,
    \`  const result = await db.\${model}.create({\`,
    '    data: req.body,',
    '  })',
    '  res.send(result)',
    '}',
    '',
    \`export default convention(handler)\`,
    '',
  ]
  return [...imps, '', ...codes].join('\\\\n')
})",
    "filename": "genCreateCode.mjs",
  },
  {
    "code": "export default makeTemplate(
  ({ Filename: Model, relative2api, filename: model }) => {
  const imps = [
    \`import type { AuthedHandler } from '\${relative2api}/../types/handler.js'\`,
    \`import type { Pure } from '\${relative2api}/../types/pure.js'\`,
    \`import { findManyArgsZ } from '\${relative2api}/validate/findManyArgs.js'\`,
    \`import type { PageResult, RawFindManyArgs } from '\${relative2api}/../types/common.js'\`,
    \`import { convention } from '\${relative2api}/convention.js'\`,
    \`import { db } from '\${relative2api}/db/prisma.js'\`,
  ]
  const codes = [
    \`type ApiType = AuthedHandler<{ query: RawFindManyArgs }, PageResult<Pure.\${Model}>>\`,
    '',
    \`/** Get page datas from \${Model} */\`,
    \`const handler: ApiType = async (req, res) => {\`,
    \`  const findManyArgs = findManyArgsZ.parse(req.query)\`,
    '  const [total, datas] = await db.$transaction([',
    \`    db.\${model}.count({ where: findManyArgs.where }),\`,
    \`    db.\${model}.findMany(findManyArgs),\`,
    '  ])',
    '  res.send({ total, datas })',
    '}',
    '',
    \`export default convention(handler)\`,
    '',
  ]
  return [...imps, '', ...codes].join('\\\\n')
})",
    "filename": "genReadCode.mjs",
  },
  {
    "code": "export default makeTemplate(({ Dir: Model, relative2api, dir: model }) => {
  const imps = [
    \`import type { AuthedHandler } from '\${relative2api}/../types/handler.js'\`,
    \`import type { Pure } from '\${relative2api}/../types/pure.js'\`,
    \`import { convention } from '\${relative2api}/convention.js'\`,
    \`import { db } from '\${relative2api}/db/prisma.js'\`,
    \`import type { RawId } from '\${relative2api}/../types/common.js'\`,
    \`import { modelIdZ } from '\${relative2api}/validate/modelId.js'\`,
  ]
  const codes = [
    \`type ApiType = AuthedHandler<\`,
    \`  { params: RawId; body: Pure.\${Model}UpdateInput },\`,
    \`  Pure.\${Model}\`,
    \`>\`,
    '',
    \`/** Update \${Model} by id */\`,
    \`const handler: ApiType = async (req, res) => {\`,
    '  const { id } = modelIdZ.parse(req.params)',
    '  const data = req.body',
    \`  const result = await db.\${model}.update({\`,
    '    where: { id },',
    '    data,',
    '  })',
    '  res.send(result)',
    '}',
    '',
    \`export default convention(handler)\`,
    '',
  ]
  return [...imps, '', ...codes].join('\\\\n')
})",
    "filename": "genUpdateCode.mjs",
  },
  {
    "code": "export default makeTemplate(({ relative2api, Dir: Model, dir: model }) => {
  const imps = [
    \`import type { AuthedHandler } from '\${relative2api}/../types/handler.js'\`,
    \`import type { RawId } from '\${relative2api}/../types/common.js'\`,
    \`import { modelIdZ } from '\${relative2api}/validate/modelId.js'\`,
    \`import { convention } from '\${relative2api}/convention.js'\`,
    \`import { db } from '\${relative2api}/db/prisma.js'\`,
  ]
  const codes = [
  \`type ApiType = AuthedHandler<{ params: RawId }>\`,
    '',
    \`/** Delete \${Model} by id */\`,
    \`const handler: ApiType = async (req, res) => {\`,
    '  const { id } = modelIdZ.parse(req.params)',
    \`  await db.\${model}.delete({ where: { id } })\`,
    '  res.send()',
    '}',
    '',
    \`export default convention(handler)\`,
    '',
  ]
  return [...imps, '', ...codes].join('\\\\n')
})",
    "filename": "genDeleteCode.mjs",
  },
  {
    "code": "// @ts-check
import genCreateCode from './genCreateCode.mjs'
import genUpdateCode from './genUpdateCode.mjs'
import genDeleteCode from './genDeleteCode.mjs'
import genReadCode from './genReadCode.mjs'
import genBasicCode from './genBasicCode.mjs'
import { defineApiTemplates } from 'tealina'

export default defineApiTemplates([
  {
    alias: 'c',
    name: '',
    method: 'post',
    generateFn: genCreateCode,
  },
  {
    alias: 'r',
    name: '',
    method: 'get',
    generateFn: genReadCode,
  },
  {
    alias: 'u',
    name: '[id]',
    method: 'put',
    generateFn: genUpdateCode,
  },
  {
    alias: 'd',
    name: '[id]',
    method: 'delete',
    generateFn: genDeleteCode,
  },
  {
    alias: '*', //fallback
    name: '',
    method: 'post',
    generateFn: genBasicCode,
  },
])",
    "filename": "index.mjs",
  },
]
`;
